addons/binding/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/connectors/ConnectorBase.java

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright (c) 2010-2018 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.smartmeter.connectors;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.reactivestreams.Publisher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.reactivex.BackpressureStrategy;
+import io.reactivex.Flowable;
+import io.reactivex.FlowableEmitter;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ * Represents a basic implementation of a SML device connector.
+ *
+ * @author Mathias Gilhuber
+ * @since 1.7.0
+ */
+@NonNullByDefault
+public abstract class ConnectorBase<T> implements IMeterReaderConnector<T> {
+
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+    /**
+     * The name of the port where the device is connected as defined in openHAB configuration.
+     */
+    private String portName;
+    private static final int NUMBER_OF_RETRIES = 3;
+
+    /**
+     * Contructor for basic members.
+     *
+     * This constructor has to be called from derived classes!
+     *
+     * @throws IOException
```

Comment:
Please go through ALL the code and add a short JavaDoc description to all `@param`, `@return` and `@throws` lines.